File: csharp/ClaudeUsageTracker/Autostart.cs
// Manage the Windows autostart entry (HKCU Run key).
//
// The registry entry is synced to the `autostart` config value on every app
// start, so a moved EXE re-registers itself with the fresh path automatically.
// The value name differs from the Python variant's ("ClaudeUsageTracker") so
// both variants can be autostarted independently.

using Microsoft.Win32;

namespace ClaudeUsageTracker;

public static class Autostart
{
    private const string RunKey = @"Software\Microsoft\Windows\CurrentVersion\Run";
    private const string ValueName = "ClaudeUsageTrackerCS";

    /// <summary>Make the HKCU Run entry match <paramref name="enabled"/>. Never throws.</summary>
    public static void Sync(bool enabled)
    {
        var exePath = Environment.ProcessPath;
        if (enabled && string.IsNullOrEmpty(exePath))
        {
            Log.Warning("autostart", "Could not determine the executable path; autostart not registered.");
            return;
        }

        try
        {
            using var key = Registry.CurrentUser.OpenSubKey(RunKey, writable: true);
            if (key is null)
            {
                Log.Warning("autostart", "Could not open the HKCU Run key.");
                return;
            }
            if (enabled)
            {
                key.SetValue(ValueName, $"\"{exePath}\"", RegistryValueKind.String);
                Log.Info("autostart", $"Autostart entry set: {exePath}");
            }
            else if (key.GetValue(ValueName) is not null)
            {
                key.DeleteValue(ValueName, throwOnMissingValue: false);
                Log.Info("autostart", "Autostart entry removed.");
            }
        }
        catch (Exception exc)
        {
            Log.Warning("autostart", $"Could not update autostart registry entry: {exc.Message}");
        }
    }
}

File: python/src/claude_usage_monitor/autostart.py
"""Manage the Windows autostart entry (HKCU Run key).

The registry entry is synced to the `autostart` config value on every app
start, so a moved EXE re-registers itself with the fresh path automatically.
"""

from __future__ import annotations

import logging
import sys
import winreg

logger = logging.getLogger(__name__)

_RUN_KEY = r"Software\Microsoft\Windows\CurrentVersion\Run"
_VALUE_NAME = "ClaudeUsageTracker"


def sync_autostart(enabled: bool) -> None:
    """Make the HKCU Run entry match *enabled*. Never raises.

    Only the packaged EXE registers itself — a python.exe/venv path would
    break as soon as the checkout moves, so source runs just log a hint and
    leave any existing EXE registration untouched.
    """
    frozen = bool(getattr(sys, "frozen", False))
    if enabled and not frozen:
        logger.warning(
            "autostart = true has no effect when running from source; "
            "build the EXE (build_exe.bat) and run that instead."
        )
        return

    try:
        with winreg.OpenKey(
            winreg.HKEY_CURRENT_USER, _RUN_KEY, 0, winreg.KEY_SET_VALUE
        ) as key:
            if enabled:
                winreg.SetValueEx(
                    key, _VALUE_NAME, 0, winreg.REG_SZ, f'"{sys.executable}"'
                )
                logger.info("Autostart entry set: %s", sys.executable)
            else:
                try:
                    winreg.DeleteValue(key, _VALUE_NAME)
                    logger.info("Autostart entry removed.")
                except FileNotFoundError:
                    pass  # was never registered — nothing to do
    except OSError as exc:
        logger.warning("Could not update autostart registry entry: %s", exc)
